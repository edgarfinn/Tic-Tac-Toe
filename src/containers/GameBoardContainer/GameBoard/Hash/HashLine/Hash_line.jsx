import React from 'react';
import styled, { keyframes } from 'styled-components';
import PropTypes from 'prop-types';

const offsetVal = ({ position }) => {
  switch (position) {
    case 'left':
    case 'top': return 'calc(33.33% - 2px);';
    case 'right':
    case 'bottom': return 'calc(66.66% - 2px);';
    default:
      return 'auto;';
  }
};

const offsetParam = ({ position }) => {
  switch (position) {
    case 'top':
    case 'bottom': return 'top';
    case 'left':
    case 'right': return 'left';
    default:
      return 'auto;';
  }
};

const HashLine = ({ linePosition, transition }) => {
  const axis = (linePosition === 'top' || linePosition === 'bottom') ? 'X' : 'Y';
  const animationParam = axis === 'X' ? 'width' : 'height';

  const lengthen = keyframes`
    0% {
      ${animationParam}: 0;
    }
    100% {
      ${animationParam}: calc(100% - 20px);;
    }
  `;

  const animationDelay = () => `${Number(transition) * 0.6}s`;

  const initialHeight = axis === 'X' ? '4px' : '0';
  const initialWidth = axis === 'Y' ? '4px' : '0';

  const StyledLine = styled.div`
    position: absolute;
    box-sizing: border-box;
    height: ${initialHeight};
    width: ${initialWidth};
    background-color: black;
    border-radius: 3px;
    ${offsetParam}: ${offsetVal};
    animation: ${lengthen} 0.7s;
    animation-fill-mode: forwards;
    animation-delay: ${animationDelay};
    animation-timing-function: cubic-bezier(.91,-0.1,.35,1);
  `;

  return (
    <StyledLine position={linePosition} />
  );
};

HashLine.propTypes = {
  linePosition: PropTypes.oneOf(['top', 'bottom', 'left', 'right']).isRequired,
  transition: PropTypes.oneOf(['1', '2', '3', '4']).isRequired,
};

export default HashLine;
