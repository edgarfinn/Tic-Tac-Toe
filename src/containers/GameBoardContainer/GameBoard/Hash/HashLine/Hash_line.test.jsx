/* eslint-env jest */
import React from 'react';
import 'jest-styled-components';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme, { mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import HashLine from './Hash_line';

Enzyme.configure({ adapter: new Adapter() });

const setup = (props) => {
  const wrapper = mount(<HashLine {...props} />);
  return { props, wrapper };
};

describe('<HashLine position="top" />', () => {
  const properties = { linePosition: 'top', transition: '3' };
  const { wrapper } = setup(properties);

  it('renders correctly position: top', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    expect(wrapper).toHaveStyleRule('top', 'calc(33.33% - 2px)');
  });
});

describe('<HashLine position="left" />', () => {
  const properties = { linePosition: 'left', transition: '1' };
  const { wrapper } = setup(properties);

  it('renders correctly position: left', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    expect(wrapper).toHaveStyleRule('left', 'calc(33.33% - 2px)');
  });
});

describe('<HashLine position="right" />', () => {
  const properties = { linePosition: 'right', transition: '4' };
  const { wrapper } = setup(properties);

  it('renders correctly position: right', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    expect(wrapper).toHaveStyleRule('left', 'calc(66.66% - 2px)');
  });
});

describe('<HashLine position="bottom" />', () => {
  const properties = { linePosition: 'bottom', transition: '2' };
  const { wrapper } = setup(properties);

  it('renders correctly position: bottom', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
