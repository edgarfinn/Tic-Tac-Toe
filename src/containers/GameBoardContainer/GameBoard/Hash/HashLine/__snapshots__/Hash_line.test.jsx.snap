// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HashLine position="bottom" /> renders correctly position: bottom 1`] = `
.c0 {
  position: absolute;
  box-sizing: border-box;
  height: 4px;
  width: 0;
  background-color: black;
  border-radius: 3px;
  top: calc(66.66% - 2px);
  -webkit-animation: eOsxFE 0.7s;
  animation: eOsxFE 0.7s;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
  -webkit-animation-delay: 1.2s;
  animation-delay: 1.2s;
  -webkit-animation-timing-function: cubic-bezier(.91,-0.1,.35,1);
  animation-timing-function: cubic-bezier(.91,-0.1,.35,1);
}

<HashLine
  linePosition="bottom"
  transition="2"
>
  <styled.div
    position="bottom"
  >
    <div
      className="c0"
    />
  </styled.div>
</HashLine>
`;

exports[`<HashLine position="left" /> renders correctly position: left 1`] = `
.c0 {
  position: absolute;
  box-sizing: border-box;
  height: 0;
  width: 4px;
  background-color: black;
  border-radius: 3px;
  left: calc(33.33% - 2px);
  -webkit-animation: xGUsi 0.7s;
  animation: xGUsi 0.7s;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
  -webkit-animation-delay: 0.6s;
  animation-delay: 0.6s;
  -webkit-animation-timing-function: cubic-bezier(.91,-0.1,.35,1);
  animation-timing-function: cubic-bezier(.91,-0.1,.35,1);
}

<HashLine
  linePosition="left"
  transition="1"
>
  <styled.div
    position="left"
  >
    <div
      className="c0"
    />
  </styled.div>
</HashLine>
`;

exports[`<HashLine position="right" /> renders correctly position: right 1`] = `
.c0 {
  position: absolute;
  box-sizing: border-box;
  height: 0;
  width: 4px;
  background-color: black;
  border-radius: 3px;
  left: calc(66.66% - 2px);
  -webkit-animation: xGUsi 0.7s;
  animation: xGUsi 0.7s;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
  -webkit-animation-delay: 2.4s;
  animation-delay: 2.4s;
  -webkit-animation-timing-function: cubic-bezier(.91,-0.1,.35,1);
  animation-timing-function: cubic-bezier(.91,-0.1,.35,1);
}

<HashLine
  linePosition="right"
  transition="4"
>
  <styled.div
    position="right"
  >
    <div
      className="c0"
    />
  </styled.div>
</HashLine>
`;

exports[`<HashLine position="top" /> renders correctly position: top 1`] = `
.c0 {
  position: absolute;
  box-sizing: border-box;
  height: 4px;
  width: 0;
  background-color: black;
  border-radius: 3px;
  top: calc(33.33% - 2px);
  -webkit-animation: eOsxFE 0.7s;
  animation: eOsxFE 0.7s;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
  -webkit-animation-delay: 1.7999999999999998s;
  animation-delay: 1.7999999999999998s;
  -webkit-animation-timing-function: cubic-bezier(.91,-0.1,.35,1);
  animation-timing-function: cubic-bezier(.91,-0.1,.35,1);
}

<HashLine
  linePosition="top"
  transition="3"
>
  <styled.div
    position="top"
  >
    <div
      className="c0"
    />
  </styled.div>
</HashLine>
`;
